File: Flutterapp/expansetracker/web/src/services/socketService.js
import io from 'socket.io-client';

class SocketService {
  constructor() {
    this.socket = null;
    this.isConnected = false;
    this.userId = null;
    this.userType = null;
    
    // Update this to your server IP
    this.serverUrl = 'http://192.168.1.8:5000'; // Change to your network IP
  }

  // Initialize socket connection
  init(userId, userType, token) {
    this.userId = userId;
    this.userType = userType;

    this.socket = io(this.serverUrl, {
      transports: ['websocket'],
      autoConnect: true,
      reconnection: true,
      reconnectionAttempts: 5,
      reconnectionDelay: 1000,
    });

    this.socket.on('connect', () => {
      console.log('🔌 Connected to server');
      this.isConnected = true;
      
      // Join user's personal room
      this.socket.emit('join_user', {
        userId,
        userType,
        name: 'User Name' // Get from user data
      });
    });

    this.socket.on('disconnect', () => {
      console.log('🔌 Disconnected from server');
      this.isConnected = false;
    });

    this.socket.on('connect_error', (error) => {
      console.error('🔌 Connection error:', error);
      this.isConnected = false;
    });

    return this.socket;
  }

  // Join a specific chat room
  joinChatRoom(roomId) {
    if (this.socket && this.isConnected) {
      this.socket.emit('join_chat_room', {
        roomId,
        userId: this.userId,
      });
    }
  }

  // Leave a chat room
  leaveChatRoom(roomId) {
    if (this.socket && this.isConnected) {
      this.socket.emit('leave_chat_room', {
        roomId,
        userId: this.userId,
      });
    }
  }

  // Send a message
  sendMessage({ roomId, senderId, receiverId, message, messageType = 'text', mediaUrl }) {
    if (this.socket && this.isConnected) {
      this.socket.emit('send_message', {
        roomId,
        senderId,
        receiverId,
        message,
        messageType,
        mediaUrl,
      });
    }
  }

  // Send typing indicator
  startTyping(roomId, userName) {
    if (this.socket && this.isConnected) {
      this.socket.emit('typing_start', {
        roomId,
        userId: this.userId,
        userName,
      });
    }
  }

  // Stop typing indicator
  stopTyping(roomId) {
    if (this.socket && this.isConnected) {
      this.socket.emit('typing_stop', {
        roomId,
        userId: this.userId,
      });
    }
  }

  // Mark messages as read
  markMessagesAsRead(roomId, messageIds) {
    if (this.socket && this.isConnected) {
      this.socket.emit('mark_messages_read', {
        roomId,
        userId: this.userId,
        messageIds,
      });
    }
  }

  // Event listeners
  onNewMessage(callback) {
    if (this.socket) {
      this.socket.on('new_message', callback);
    }
  }

  onMessageNotification(callback) {
    if (this.socket) {
      this.socket.on('message_notification', callback);
    }
  }

  onUserTyping(callback) {
    if (this.socket) {
      this.socket.on('user_typing', callback);
    }
  }

  onUserStoppedTyping(callback) {
    if (this.socket) {
      this.socket.on('user_stopped_typing', callback);
    }
  }

  onMessagesRead(callback) {
    if (this.socket) {
      this.socket.on('messages_read', callback);
    }
  }

  onUserStatusChanged(callback) {
    if (this.socket) {
      this.socket.on('user_status_changed', callback);
    }
  }

  // Disconnect
  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.isConnected = false;
    }
  }
}

export default new SocketService();

File: Flutterapp/expansetracker/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  
  IO.Socket? _socket;
  bool get isConnected => _socket?.connected ?? false;
  IO.Socket get socket => _socket!;
  
  // Use your second laptop's IP where MERN server is running
  // Replace with your second laptop's actual IP address
  //static const String serverUrl = 'http://192.168.65.182:5001'; // Update this IP
  static  String serverUrl = 'http://192.168.1.11:5001';
  
  SocketService._internal();

  // Initialize socket connection
  Future<bool> initSocket(String userId, String userType, String authToken) async {
    try {
      print('🔌 SocketService: Initializing socket for user $userId ($userType)');
      print('🌐 Connecting to server: $serverUrl');
      
      // Disconnect existing socket if any
      if (_socket != null) {
        _socket!.disconnect();
        _socket = null;
      }
      
      // Create new socket connection with timeout
      _socket = IO.io(
        serverUrl,
        IO.OptionBuilder()
          .setTransports(['websocket', 'polling'])
          .enableAutoConnect()
          .enableReconnection()
          .setReconnectionAttempts(3) // Reduced attempts
          .setReconnectionDelay(2000)
          .setTimeout(10000) // 10 second timeout
          .setQuery({
            'userId': userId,
            'userType': userType,
            'token': authToken,
            'from': 'flutter'
          })
          .build(),
      );

      bool connected = false;
      
      _socket!.onConnect((_) {
        print('✅ Connected to MERN server: ${_socket!.id}');
        connected = true;
        
        // Send authentication data matching MERN format
        _socket!.emit('authenticate', {
          'userId': userId,
          'userType': userType,
          'name': userType == 'patient' ? 'Patient User' : 'Doctor User',
        });
      });

      _socket!.onDisconnect((_) {
        print('❌ Disconnected from server');
        connected = false;
      });
      
      _socket!.onError((error) {
        print('❌ Socket error: $error');
        connected = false;
      });

      _socket!.onConnectError((error) {
        print('❌ Connection error: $error');
        connected = false;
      });

      // Wait for connection with timeout
      int attempts = 0;
      while (!connected && attempts < 10) {
        await Future.delayed(Duration(milliseconds: 500));
        attempts++;
        if (_socket!.connected) {
          connected = true;
          break;
        }
      }
      
      if (!connected) {
        print('❌ Failed to connect after timeout');
        _socket?.disconnect();
        _socket = null;
      }
      
      return connected;
      
    } catch (e) {
      print('❌ SocketService initSocket error: $e');
      return false;
    }
  }

  // Join a chat room
  Future<bool> joinRoom(String roomId, String userId) async {
    try {
      if (_socket?.connected == true) {
        _socket!.emit('join_room', {
          'roomId': roomId,
          'userId': userId,
        });
        print('🏠 Joined room: $roomId');
        return true;
      } else {
        print('❌ Cannot join room: Socket not connected');
        return false;
      }
    } catch (e) {
      print('❌ Error joining room: $e');
      return false;
    }
  }

  // Send a message
  void sendMessage(String roomId, String message, String senderId) {
    if (_socket?.connected == true) {
      _socket!.emit('send_message', {
        'roomId': roomId,
        'senderId': senderId,
        'receiverId': '', // Will be determined by room
        'text': message,
        'type': 'text',
        'timestamp': DateTime.now().toIso8601String(),
      });
      print('📤 Message sent: $message');
    }
  }

  // Listen for new messages
  void listenForMessages(Function(dynamic) onMessageReceived) {
    if (_socket != null) {
      _socket!.on('receive_message', (data) {
        print('📨 Received message: $data');
        onMessageReceived(data);
      });
    }
  }

  // Send typing indicator
  void emitTyping(String roomId, String userId) {
    if (_socket?.connected == true) {
      _socket!.emit('typing_start', {
        'roomId': roomId,
        'userId': userId,
        'userName': 'User',
      });
    }
  }

  // Stop typing indicator
  void stopTyping(String roomId, String userId) {
    if (_socket?.connected == true) {
      _socket!.emit('typing_stop', {
        'roomId': roomId,
        'userId': userId,
      });
    }
  }

  // Disconnect
  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket = null;
    }
  }
}